README.md
# JspitTemplate - Smart PHP Template Class

JspitTemplate is a small and fast template system for PHP.

## Features
- pure HTML templates 
- integrated escaping
- Filters for formatting, escaping, user defined 
- small, only one file with 15kByte
- Independent (Linux,Window, PHP >= V7.0)

## How to use it

Easy, just download and include the class file and and make outputs wherever you want.

### Simple example

index.tpl.html
```html
<!DOCTYPE html>
<html lang="de">
  <head>
    <meta charset="UTF-8" />
    <title>{{ title ?? 'Demo JspitTemplate'}}</title>
  </head>
  <body>
    <h3>{{ h3 }}</h3>
    {{content}}
  </body>
</html>
```

demo_jspittemplate_simple.php
```php
require __DIR__.'/JspitTemplate.php';

header('Content-Type: text/html; charset=UTF-8');

echo JspitTemplate::create(__DIR__.'/index.tpl.html')
  ->assign([
      'h3' => 'JspitTemplate Simple Test',
      'content' => '<Hallo World>',
   ])
;
```

HTML:
```html
<!DOCTYPE html>
<html lang="de">
  <head>
    <meta charset="UTF-8" />
    <title>Demo JspitTemplate</title>
  </head>
  <body>
    <h3>JspitTemplate Simple Test</h3>
    &lt;Hallo World&gt;
  </body>
</html>
```

## Placeholder

A simple placeholder is just a name within two curly brackets.

```html
{{ h3 }}
```
The placeholders are replaced by application variables using the assign method with an array as a parameter. The array consists of one or more elements of the form 'placeholder' => 'value'.

```php
  ->assign([
     'h3' => 'JspitTemplate Simple Test',
   ])
```

Placeholder may have attributes or elements. 
Use a dot (.) to access attributes of a variable (items of a PHP array or properties of a PHP object).

```html
{{ foo.bar }}
```

Behind ?? optional default values can be used. A default value must be a fixed string. This string is always escaped.

```html
{{ foo.bar ?? 'default' }}
```

Filters can be placed behind the placeholders. Filters are marked by a | .

```html
{{ foo.bar|filter ?? 'default' }}
```

## Filters

Placeholder can be modified by filters. Filters are separated from the placeholder by a pipe symbol ( | ). Multiple filters can be chained. The output of one filter is applied to the next.

### raw

The raw filter means that this variable is not masked. Without a raw filter, the variables are masked with htmlspecialchars().
If the variable contains html that should be interpreted by the browser, the filter html (alias for raw) or raw must be used.

```html
{{ htmlcontent|raw }}
```

### url

The url filter encodes a given string as URL segment with rawurlencode() or an array as query string with http_build_query().

```html
{{ par|url }}
```

### format

The format filter returns the assigned variable like sprintf()/vsprintf().

```html
{{ var|format('%03d') }}
```

### date

Formats a string, an object from the DateTimeInterface or an integer time stamp to a date.

```html
{{date|date(" d.m.Y")}}
```

### case

The filter case selects a string that corresponds to the value converted to an integer. The individual strings are separated by the first character in the chain.

```html
{{ var|case('#zero#one#two#') }}
```

If the value 1 is assigned to var in the example, the expression 'one' results. A empty string is the result if there is no corresponding case.

### set

If var is set and assigned and not NULL, the value is replaced with the fixed string used as the parameter for set.

```html
{{var|set('var is set') ?? 'var not set'}}
```

### blank

The filter blank does the same as set("").

### selected

selected is a special filter for forms. It enables the selected property to be set to restore the selection after the form has been sent. The standard value must always be set here.

```html
{{post.sel|selected('p2') ?? ' '}}
```

The result is "selected" if the element "sel" from the array post is equal to "p2". For all other cases the result is an empty string. The standard value must always be set here.

### checked

checked is a special filter for checkboxes in forms. 

```html
{{post.chk|checked ?? ''}}
```

The result is "checked" if the "chk" element is present in the post array. In all other cases the result is an empty string.

Custom filters can be added using the addStaticUserFunction() and addUserFunction() methods. Filters that are not defined always return the unchanged value and do not cause any errors.

### each

Simple loops can be implemented with the filter 'each'.

```html
<ul>
  {{htmlList|each("<li>#val#</li>")|raw}}
</ul>
```

htmlList must be assigned an array. The placeholder #val# is replaced with the value and #key' with the key from the array.

```php
$html = JspitTemplate::createFromString($template)
  ->render(['htmlList' => ['Coffee','Tea','Milk']])
;
```

HTML:
```html
<ul>
<li>Coffee</li>
<li>Tea</li>
<li>Milk</li>
</ul>
```



## Doc
http://jspit.de/tools/classdoc.php?class=JspitTemplate 

## Demo and Test
http://jspit.de/check/phpcheck.jspittemplate.php




changelog.md
# JspitTemplate

## Version 1.5 (2022-07-19)
* Added Filter 'each' for simple loops

## Version 1.4 (2021-12-04)
* Added Filter sign and abs

## Version 1.3 (2021-07-29)
* Added Filter case

## Version 1.2 (2021-06-21)
* Added method assignOnce

## Version 1.1 (2021-01-12)
first Version 


JspitTemplate.php
<?php
//namespace Jspit;
/**
	JspitTemplate.php
.---------------------------------------------------------------------------.
|  Software: JspitTemplate - Simple PHP Template Class                      |
|   Version: 1.5                                                            |
|      Date: 2022-07-19                                                     |
|  PHPVersion >= 7.0                                                        |
| ------------------------------------------------------------------------- |
| Copyright © 2020 - 2021, Peter Junk (alias jspit). All Rights Reserved.   |
| ------------------------------------------------------------------------- |
|   License: Distributed under the Lesser General Public License (LGPL)     |
|            http://www.gnu.org/copyleft/lesser.html                        |
| This program is distributed in the hope that it will be useful - WITHOUT  |
| ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or     |
| FITNESS FOR A PARTICULAR PURPOSE.                                         |
'---------------------------------------------------------------------------'
 */

class JspitTemplate {

  protected static $templatePath = "" ;
  protected static $staticUserFct = [];
  private $tpl = "";  //Template Content
  private $assign = [];
  private $userFct = [];

  /*
   * @param $tplFileName Template-Filename or Template-String
   * @param $flagAsString true if $tplFileName is a template-string, default false
   * @throws Exception if Template-File is not readable
   */
  public function __construct(string $tplFileName = NULL, bool $flagAsString = false) 
  {
    if($tplFileName != NULL){
      if($flagAsString) {
        $this->loadString($tplFileName);
      }
      else {
        $this->loadFile($tplFileName);
      }
      $this->userFct = self::$staticUserFct;
    }
  }

  /*
   * create a object JspitTemplate
   * @param $tplFileName Template-Filename
   * @return object or throw a exeption if error
   */
  public static function create(string $tplFileName) : self
  {
    try{
      $tpl = new self($tplFileName);
    }
    catch (\Exception $e) {
      throw new \InvalidArgumentException('Error read Templatefile "'.$tplFileName.'"');
    }
    return $tpl;
  }
  
  /*
   * createFromString
   * create a object JspitTemplate from a Template-String
   * @param $string a Template
   * @return object self
   */
  public static function createFromString(string $string = "") : self
  {
    return new self($string, true);
  }

  /*
   * renderFromString
   * render a Template-String
   * @param $string a Template
   * @param array $keyValue 
   * @return string HTML
   */
  public static function renderFromString(string $string = "",array $assign = []) : string
  {
    return (new self($string, true))->render($assign);
  }


 /*
  * set default template path
  * @param string $path, without $path or "" reset template path
  * @return realpath if $path exists or false if not
  */
  public static function setTemplatePath(string $path = '') : string
  {
    if($path != ''){
      $path = str_replace('\\','/',$path);
      if(substr($path,-1) != '/') $path .= '/';
    }
    self::$templatePath = $path;
    return $path;
  }

/*
 * render template and get html
 * @param string filename (optional)
 * @param array $keyValue 
 * @return string HTML
 * @throw Exception if arguments incorrect
 */
  public function render(...$par)
  {
    if(isset($par[0])){
      if(is_string($par[0])){
        $this->loadFile(array_shift($par));
      }
      if(is_array($par[0])) {
        return $this->getHTMl($par[0]);
      }
      else {
        throw new \InvalidArgumentException("Error Arguments render");
      } 
    }
    else {
      return $this->getHTMl();
    } 
  }

 /*
  * render code and get HTML
  * @param array $keyValue 
  * @return string HTML
  */ 
  public function getHTML(array $assign = []) : string
  { 
    if(!empty($assign)) $this->assignOnce($assign);
    $this->assignOnce($this->assign);
    $this->replaceDefaults();
    return $this->tpl;
  }

 /*
  * magic function, please refer getHTML
  * @return string HTML
  */
  public function __toString() : string
  {
    return $this->getHTML();
  }

 /*
  * load Template
  * @param String $tplFileName : Filename 
  * @return $this
  * @throw Exception if Template-FileName incorrect
  */
  public function loadFile(string $tplFileName) : self
  {
    if(!preg_match('~^/|^\\\\|^[A-Z]:~',$tplFileName)) {
      //relative Path
      $tplFileName = self::$templatePath.$tplFileName;
    }
    if(!is_readable($tplFileName)) {
      throw new \Exception("Error: Template-File '".$tplFileName."' is not readable");
    }
    $this->tpl = file_get_contents($tplFileName);
    $this->prepare();
    return $this;
  }

 /*
  * load String
  * @param string Template
  * @return $this
  */
  public function loadString(string $template) : self
  {
    $this->tpl = $template;
    //check and prepare
    $this->prepare();
    return $this;
  }


 /*
  * get Template
  * @return string Template 
  */
  public function getTemplate() : string
  {
    return $this->tpl;
  }


 /*
  * save Template
  * @param string $tplFileName
  * @return $this object 
  */
  public function save(string $tplFileName) : self
  {
    if(!preg_match('~^/|^\\\\|^[A-Z]:~',$tplFileName)) {
      //relative Path
      $tplFileName = self::$templatePath.$tplFileName;
    }
    file_put_contents($tplFileName, $this->tpl);
    return $this;
  }

 /*
  * assign : Assign content to placeholders 
  * @param $keyValue array, object or null (default)
  * @return $this object 
  */
  public function assign($keyValue = NULL) : self 
  {
    if(is_object($keyValue) or is_array($keyValue)) {
      $set = (array)$keyValue;
      $this->assign = array_replace_recursive($this->assign, $set);
    }
    elseif($keyValue === NULL){
      //delete all assigns
      $this->assign = [];
    }
    else {
      //error
      $msg = "Invalid first parameter for ".__METHOD__;
      throw new \InvalidArgumentException($msg);
    }
    return $this;
  }

 /*
  * assignOnce : Assign content to placeholders at once
  * @param $keyValue array or object
  * @param bool $insert, default false, true insert before the placeholder
  * @return $this object 
  */
  public function assignOnce($keyValue, bool $insert = false) : self 
  {
    if(is_object($keyValue) or is_array($keyValue)) {
      $set = (array)$keyValue;
    }
    else {
      //error
      $msg = "Invalid first parameter for ".__METHOD__;
      throw new \InvalidArgumentException($msg);
    }
    foreach($set as $key => $val){
      if($val instanceOf self){
        $tpl = $val->assignOnce($val->assign)->getTemplate();
        $this->replaceplaceholder($key, $tpl);
      }
      else {
        foreach($this->getPlaceholders($key) as $tplPlaceholder){
          $fraction = self::splitPlaceholder($tplPlaceholder);
          if($fraction === false OR $val === []) continue;
          $att = $fraction['name'][1] ?? null;
          if($att !== null){  //array or object
            if(is_array($val) AND isset($val[$att])) $newVal = $val[$att];
            elseif(is_object($val) AND isset($val->$att)) $newVal = $val->$att;
            else continue; 
          }
          else {
            $newVal = $val;
          }
          //filters
          if($newVal !== NULL){
            $newVal = $this->handleFilter($fraction['filter'], $newVal);
            //replace
            if($insert) $newVal .= $tplPlaceholder;
            $this->tpl = str_replace($tplPlaceholder, $newVal, $this->tpl);
          }  
        }
      }
    }
    return $this;
  }

 /*
  * getPlaceholders : use internal and for test
  * @param $name name or "" for all
  * @return array("{{foo|format(\"%3d\")}}",..)
  */
  public function getPlaceholders(string $name = "") : array
  {
    $search = $name !== ""  
      ? (' ?'.preg_quote($name,'~').'([. |?][^}]*)?')
      : ('[^}]+')
    ;
    $regEx = '~\{\{'.$search.'\}\}~u';
    preg_match_all($regEx, $this->tpl, $matches);
    return array_unique($matches[0]);
  }

 /*
  * add a userfuction (closure)
  * @param string $name
  * @param $function : closure
  * @return true
  */
  public static function addStaticUserFunction(string $name, $function) : bool
  {
    self::$staticUserFct[$name] = $function;
    return true;
  }

 /*
  * add a userfuction (closure)
  * @param string $name
  * @param $function : closure
  * @return $this
  */
  public function addUserFunction(string $name, $function) : self
  {
    $this->userFct[$name] = $function;
    return $this;
  }


 /*
  * private
  */

 /*
  * check and prepare template
  */
  private function prepare() : void
  {
    $regEx = '/(?<!\\\\)([\'"])(?:(?!(?:\1|\\\\)).|\\\\.)*+\1(*SKIP)(*FAIL)|\h+/m'; 
    $regExCheck = '~^\{\{\w+(\.\w+)?(.+)?\}\}$~';
    foreach($this->getPlaceholders() as $placeHolder){
      $compress = preg_replace($regEx, '', $placeHolder);
      if(!preg_match($regExCheck, $compress)){
        //format error placeholder
        throw new \Exception("Format Error Template Placeholder '".$placeHolder."'");
      }
      $this->tpl = str_replace($placeHolder, $compress,$this->tpl);
    }
  }

 /*
  * replace all placeholders in the template with the value
  */
  private function replaceplaceholder(string $placveholder, string $value)
  {
    $placveholder = preg_quote($placveholder,'~');
    $regEx = '~\{\{ ?'.$placveholder.' ?([|?][^}]+)?\}\}~u';
    $this->tpl = preg_replace($regEx,$value, $this->tpl);
  }

 /*
  * split placeholder in komponents name, filter, default
  * @param $placeholder string e.g.'{{foo.bar|format("%3d")|html??"default val"}}'
  * @return array('name' => .., 'filter' => .., 'default' => ..) or false if error
  */ 
  public static function splitPlaceholder(string $placeholder) 
  {
    $regEx = '~^(?<name>[\w.]+)(?<filter>\|.+?)?(?<default>\?\?.+?)?$~';
    if(!preg_match($regEx, trim($placeholder,'{}'), $match)) return false;
    $fractions = [
      'name' => explode('.', $match[1]),
      'filter' => isset($match[2])
        ? explode('|',trim($match[2],'|'))
        : [],
      'default' => isset($match[3]) 
        ? self::stripQuotes(substr($match[3],2)) 
        : null 
    ];
    return $fractions;
  }

 /*
  * 'abc' => abc, "abc" => abc, 'abc" => 'abc"
  */
  public static function stripQuotes(string $str) : string
  {
    return preg_replace('/^(\'(.*)\'|"(.*)")$/', '$2$3', $str);
  }

 /*
  * replace all placeholders in the template with eascaping values
  */
  private function replaceDefaults()
  {
    $regEx = '~\{\{[^?{}]+\?\?([\'"])(.*?)(\1)\}\}~u';
    $this->tpl = preg_replace_callback(
      $regEx,
      function(array $matches){
        return htmlspecialchars($matches[2], ENT_QUOTES | ENT_SUBSTITUTE, 'UTF-8');
      },
      $this->tpl
    );
  }

 /*
  * handle filter
  * @param string $filter : array with filter strings 
  * 'filter' from placeholder e.g.  ['format("%3d")','html'] 
  * @param mixed $value
  * @return new $value
  */
  private function handleFilter(array $filter, $value) : string
  {
    $esc = true; //optional esc if not html and not url 
    foreach($filter as $fkt){
      if($fkt == "") continue;  
      //split name as $match[1] + argument as $match[2]
      $r =preg_match('~^(\w+)(?:\((.*)\))?$~', $fkt, $match);
      if($r === 0) continue;
      $filter = $match[1]; //filtername
      $arg2 = isset($match[2]) ? trim($match[2],'\'"') : false;
      if($filter == 'html' OR $filter == 'raw') $esc = false;
      elseif(substr($filter,0,3) == 'url') {
        if(is_array($value)) {
          $value = http_build_query($value);
        }
        else {
          $value = rawurlencode($value);
        }
      }
      elseif($filter == 'format' AND $arg2 !== false){
        $value = is_array($value) ? vsprintf($arg2, $value) : sprintf($arg2, $value);
      }
      elseif($filter == 'date' AND $arg2 !== false){
        if(is_numeric($value)) { //timestamp
          $tz = new \DateTimeZone(date_default_timezone_get());
          $value = date_create('@'.$value)->setTimeZone($tz);
        }
        elseif(is_string($value) AND ($dt = date_create($value)) !== false){
          $value = $dt;
        }
        if($value instanceOf \DateTimeInterface) $value = $value->format($arg2);
      }
      elseif(in_array($filter,['selected','checked'])){
        if( $arg2 !== false){
          $value = $value === $arg2 ? $filter : '';
        }
        else {
          $value = $filter; 
        }
      }
      elseif($filter == 'blank') $value = '';
      elseif($filter == 'set') {
        $value = $value !== "" ? (string)$arg2 : "";
      }
      elseif(($filter == 'case' OR $filter == 'sign') AND $arg2 !== false AND $arg2 != ""){
        $delim = substr($arg2,0,1);
        $cases = explode($delim, substr($arg2,1));
        $value = $filter == 'case' 
          ? intval($value)
          : ($value <=> 0) +1  
        ;
        $value = array_key_exists($value, $cases) 
          ? $cases[$value]
          : ""
        ;
      }
      elseif($filter == 'abs' AND is_numeric($value)) {
        $value = abs($value);
      }
      elseif($filter == 'each') {
        $value = $this->filterEach($value,$arg2);
      }
      elseif(array_key_exists($filter, $this->userFct)){
        $value = ($arg2 !== false AND $arg2 !== "") 
          ? $this->userFct[$filter]($value,$arg2)
          : $this->userFct[$filter]($value)
        ;
      }
    }   
    return $esc 
      ? htmlspecialchars((string)$value, ENT_QUOTES | ENT_SUBSTITUTE, 'UTF-8') 
      : (string)$value
    ;
  }
  
 /*
  * for each in $value 
  * @param $value (array)
  * @param string $arg  with placeholder for #key# and #val#
  * @param bool $esc escape keys and values
  * @return string new value
  */
  private function filterEach($value,$arg,$esc = true)
  {
    if(is_array($value)){
      $newValue = "";
      foreach($value as $key => $el){
        if($esc){
          $key = htmlspecialchars((string)$key, ENT_QUOTES | ENT_SUBSTITUTE, 'UTF-8');
          $el = htmlspecialchars((string)$el, ENT_QUOTES | ENT_SUBSTITUTE, 'UTF-8'); 
          $newValue .= strtr($arg, ['#key#' => $key, '#val#' => $el])."\r\n"; 
        }
      }
      return trim($newValue);
    }
    //error
    $msg = "Filter each requires an array" ;
    throw new \Exception($msg);
  }

}
